core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala

```diff
@@ -200,20 +187,52 @@ class CassandraJournal(cfg: Config) extends AsyncWriteJournal with CassandraReco
               case write :: remainder => writeMessages(write).flatMap(result => rec(remainder, result :: acc))
               case Nil                => Future.successful(acc.reverse)
             }
+
           rec(groups, Nil)
         }
 
-      result.onComplete { _ =>
+      val tagWrites = result.map({ _ =>
+        val messagesByTag: Map[String, Seq[Serialized]] = serialized
+          .flatMap(_.payload)
+          .flatMap(s => s.tags.map((_, s)))
+          .groupBy(_._1).mapValues(_.map(_._2))
+
+        messagesByTag.map {
+          case (tag, writes) =>
+            TagWrites(tag, writes.toVector)
+        }.toVector
+      })
```

Comment:
perhaps extract this to a separate method and optimize for the single event single tag case, lot's of collection juggling here